[AMD] MI350 support for ds_read_b64_tr_b16  instruction

plognjen opened the pull request:
This PR introduces initial support for using LDS transposed load intrinsics in Triton. It implements selection of ds_read_b64_tr_b16 instruction for mi350 architecture. These intrinsics are designed to handle tensors that are non-k contiguous in LDS, enabling efficient loading in a transposed manner. 

LDS transposed loading only works if:
1. The hardware supports transposed LDS load instructions.
2. The tensor we’re loading is contiguous along the non-K dimension in LDS.

How the compiler handles LDS transpose loads (like the intrinsics and LL we use) depends on two things:
1. The type of data we’re working with.
2. The type of MFMA instruction.

Currently, we only support transpose loads for 16-bit data on gfx950.
TODO: Add support for other data types (e.g., 8-bit types).
TODO : Support transpose loading on swizzled data to avoid bank conflicts.

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -30,8 +30,11 @@ namespace {
 public:
-  using ConvertOpToLLVMPattern<
-      triton::gpu::LocalLoadOp>::ConvertOpToLLVMPattern;
+  LocalLoadOpConversion(const LLVMTypeConverter &converter,
```
I think it's better to use a separate pattern with a higher benefit for the transposed case--these two cases are largely disjoint; separate patterns are more clear and avoids branchy code here.

plognjen replied:
I agree. Done.

plognjen commented on include/triton/Dialect/TritonGPU/IR/LinearLayoutConversions.h:
```diff
@@ -256,6 +256,10 @@ LinearLayout chooseStMatrixLayout(MLIRContext *ctx, RankedTensorType tensorTy,
 // tensor into shared memory using the `ldmatrix` instruction.
 LinearLayout chooseLdMatrixLayout(Attribute enc, ArrayRef<int64_t> shape,
                                   bool needTrans, int32_t elemBitWidth);
+// The primary goal of this function is to efficiently load 2D tiles of a
```
Done.

plognjen commented on third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp:
```diff
@@ -99,6 +99,15 @@ bool TargetInfo::canUseStMatrix(RankedTensorType tensorTy,
+        bitwidth == 16 || bitwidth == 8 || bitwidth == 4 || bitwidth == 6;
+    return bitwidthSupported;
+  }
+  return false;
```
Done.

antiagainst commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  // The MFMA selection logic prioritizes double-rate MFMA instructions whenever
+  // possible. Specifically:
+  // - For MFMA operations that are non-K = 16, when blockK > 16, mfma16x16x32
+  // is selected; otherwise (blockK ≤ 16), mfma16x16x32 remains the choice.
```
`mfma16x16x16`?

antiagainst commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+                           warpLayout.transposeOuts(outDimNames);
+  auto finalLayout =
+      combineCtaCgaWithShape(ctaLayout, mfmaLayout.getCTALayout(), shape);
+  // llvm::outs() << finalLayout;
```
Drop debugging code.

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,108 @@ struct LocalLoadOpConversion
+  TransLocalLoadOpConversion(const LLVMTypeConverter &converter,
+                             const AMD::TargetInfo &targetInfo,
+                             PatternBenefit benefit = 2)
+      : ConvertOpToLLVMPattern<triton::gpu::LocalLoadOp>(converter, benefit),
```
Nit: can drop `<triton::gpu::LocalLoadOp>` to make it a bit simpler.

antiagainst commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,108 @@ struct LocalLoadOpConversion
+    RankedTensorType dstTy = op.getType();
+    Attribute srcLayout = srcTy.getEncoding();
+    Attribute dstLayout = dstTy.getEncoding();
+    if (isa<DotOperandEncodingAttr>(dstLayout) &&
+        isa<AMDMfmaEncodingAttr>(
+            cast<DotOperandEncodingAttr>(dstLayout).getParent())) {
```
Nit: It's a bit simpler to write as 
```c++
if (auto dotLayout = dyn_cast<DotOperandEncodingAttr>(dstLayout))
  if (isa<AMDMfmaEncodingAttr>(dotLayout.getParent()))
```

antiagainst reviewed (changes requested):
Neat! Looks good to me; just few nits.

plognjen commented:
@antiagainst thanks for the review! Updated code according to the comments. Also I added workaround so we can have all dot test passing on mi350 before we merge double-rated mfma instructions (since ds_trans_read lowering logic assumes double-rate mfma). It is explained in the comment here:
https://github.com/triton-lang/triton/pull/5750/files#diff-64c4572b14c29aa0f0c5ef8a337e304f92cc6c7f5acb0c56715d5bdc68b22f63R145

antiagainst reviewed (approved):
LGTM now. But I'll wait for @zhanglx13 to take another look before landing given the amount of details here.

zhanglx13 commented:
I'll take a look later today or tomorrow.

zhanglx13 commented on third_party/amd/lib/TritonAMDGPUToLLVM/TargetInfo.cpp:
```diff
@@ -99,6 +99,10 @@ bool TargetInfo::canUseStMatrix(RankedTensorType tensorTy,
 }
 
+bool TargetInfo::canUseLDSTransLoad(int bitwidth) const {
+  return arch == "gfx950" && llvm::is_contained({16, 8, 4, 6}, bitwidth);
```
Since this PR only enabled bitwidth = 16, we need to respect the limitation here. And add a TODO saying that we will support other bitwidth in future PRs.

zhanglx13 commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,129 @@ struct LocalLoadOpConversion
+  bool canUseTransLoad(MemDescType srcTy, RankedTensorType dstTy) const {
+    auto bitwidth = typeConverter->convertType(dstTy.getElementType())
+                        .getIntOrFloatBitWidth();
+
```
This function checks a number of conditions to use ds_read_b64_tr_b16 instruction. It will be more organized if you can group the conditions in the following categories and check them in the following order:
1. GPU arch and bitWidth. This is done by `targetInfo.canUseLDSTransLoad(bitwidth)`
2. Correctness/Validity:
    - dotOperandLayout's parent must be mfma layout. I think we should absorb the checks at line 121 and 122 into this function
    - LDS layout. This checks nonKContig
3. Performance. Check kWidth == expectedKWidth
4. Current limitations. Check bitWidth == 16

zhanglx13 commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,129 @@ struct LocalLoadOpConversion
+      return false;
+    }
+
+    // The transposed load lowering logic assumes that double-rate MFMA
```
"double-rate MFMA" is not clear. Can you be specific about the mfma instructions you are referring to?

zhanglx13 commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,129 @@ struct LocalLoadOpConversion
+    const int32_t mDim = mfmaEnc.getMDim();
+    assert((mDim == 32 || mDim == 16) && "Invalid MFMA instruction dimension");
+
+    const int largeTileThreshold = (mDim == 32) ? 8 : 16;
```
8 and 16 here are corresponding to bitWidth = 16. For such hard-coded numbers, it's better to explain their use cases.

zhanglx13 commented on third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp:
```diff
@@ -102,10 +102,129 @@ struct LocalLoadOpConversion
+
+    auto llvmElemTy = typeConverter->convertType(dstTy.getElementType());
+    auto bitwidth = llvmElemTy.getIntOrFloatBitWidth();
+    auto ldsTransLayout = chooseLDSTransLayout(dotEnc, shape, bitwidth);
```
The names (both the variable and function names) are misleading. It sounds like a new layout for LDS. Actually, it's a regLayout, as indicated by the API of `emitTransferBetweenRegistersAndShared`. 
We can follow nv naming converntion, i.e. include the ds_read_b64_tr_b16 instruction name in the function name, such as `chooseDsReadB64Tr16Layout`. Then for the variable, we can name it as `dsReadTransLayout`.

I'd suggest we name the function as above, i.e. expose the instruction details. It's aligned with the current limitation that we only support 16-bit trans load. Later, when we support more bitWidth, we can see if it's easy to extend the function in place, or we may need add more `chooseXXX` functions for each trans load bitWidth.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  bool hasBatchDim = rank == 3;
+  int mIndex = 0 + hasBatchDim;
+  int32_t kWidthDot = dotMfmaLayout.getKWidth();
+  int32_t kWidthTransRead = 64 / elemBitWidth;
```
64 is a constant for bitWidth = 16. It's still better to define a constant variable instead of hard-coding it.

plognjen replied:
64 is actually read width for 16b, 8b and 4b elements. It is only different for 6b (128). I added a comment and variable.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  SmallVector<unsigned> order = triton::gpu::getOrder(dotMfmaLayout);
+  std::swap(order[0], order[1]);
+
+  // In the LDS transpose logic, each thread accesses 64 bits (8 bytes) of data.
```
Do we have a public documentation that can explain how LDS transpose load works?

zhanglx13 replied:
Let me ask around. If not found, we'll write one for triton.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+
+  const bool isMfma32 = (mfmaLayout.getMDim() == 32);
+  const bool isMfma16 = (mfmaLayout.getMDim() == 16);
+  assert(isMfma32 || isMfma16);
```
We already assert'ed at the beginning of this function. You may want to leave one of them.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  assert(isMfma32 || isMfma16);
+
+  const int kTileSize = isMfma32 ? 16 : 32;
+  const int kThreshold = isMfma32 ? 8 : 16;
```
`kThreshold` is not needed here. We can just do `kSize >= kTileSize` in the following branch due to block sizes are always power-of-2.

zhanglx13 reviewed (changes requested):
Nothing about functionality, just some suggestions to make things more organized.

Two things about the PR description
1. "This ensures that threads still receive k-contiguous elements in VGPRs." This is not correct. Threads will always receive k-contiguous elements in VGPR since this is the hardware requirement to use mfma instructions.
2. In the title, you need to be specific about "initial support" by providing
    - GPU arch 
    - the lds transpose load instruction name

plognjen reviewed (commented):
@zhanglx13 Code is updated. Thanks for the review.

zhanglx13 commented on lib/Dialect/TritonGPU/IR/LinearLayoutConversions.cpp:
```diff
@@ -390,6 +390,136 @@ AMDMfmaEncodingAttr::toLinearLayout(ArrayRef<int64_t> shape) const {
+  int32_t kWidthDot = dotMfmaLayout.getKWidth();
+  // Number of bits loaded by an LDS read. ds_read_tr primarily supports 64-bit
+  // loads for most element sizes (16b, 8b, 4b), except for 6b elements. Here,
+  // we assume 6b elements will not be used.
```
I'd suggest remove 6b info here. This function only handles 16-bit elemTy.

zhanglx13 reviewed (approved):
LGTM! Thanks @plognjen

zhanglx13 commented on test/Conversion/amd/ds_transpose.mlir:
```diff
@@ -0,0 +1,77 @@
+// RUN: triton-opt %s --split-input-file --convert-triton-amdgpu-to-llvm=arch=gfx950 | FileCheck %s
+
+#mma16 = #ttg.amd_mfma<{versionMajor = 3, versionMinor = 0, warpsPerCTA = [2, 2], instrShape = [16, 16], isTransposed = true}>
+#mma32 = #ttg.amd_mfma<{versionMajor = 3, versionMinor = 0, warpsPerCTA = [2, 2], instrShape = [32, 32], isTransposed = true}>
```
Does gfx950 have versionMajor = 4? Or is the change pending in another PR?

antiagainst replied:
Good point. It's added in https://github.com/triton-lang/triton/pull/5873. I just merged in `main`; tests might fail due to that.
